igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,361 @@
+    --------
+    requests.Session
+        for details of the standard options
+    """.strip()
+
+
+def _find_cred(func, *args, error=True, **kwargs):
+    """Find a credential and maybe ignore an `~igwn_auth_utils.IgwnAuthError`
+
+    This is an internal utility for the `SessionAuthMixin._init_auth`
+    method which shouldn't necessary fail if it doesn't
+    find a credential of any one type, but should just move on to the
+    next option.
+    """
+    try:
+        return func(*args, **kwargs)
+    except IgwnAuthError:
+        if error:
+            raise
+        return
+
+
+def _hook_raise_for_status(response, *args, **kwargs):
+    """Response hook to raise exception for any HTTP error (status >= 400)
+
+    Reproduced (with permission) from :mod:`requests_gracedb.errors`,
+    authored by Leo Singer.
+    """
+    return response.raise_for_status()
+
+
+class SessionErrorMixin:
+    """A mixin for :class:`requests.Session` to raise exceptions for HTTP
+    errors.
+
+    Reproduced (with permission) from :mod:`requests_gracedb.errors`,
+    authored by Leo Singer.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.hooks.setdefault("response", []).append(
+            _hook_raise_for_status,
+        )
+
+
+class SessionAuthMixin:
+    """Mixin for :class:`requests.Session` to add support for IGWN auth.
+
+    By default this mixin will automatically attempt to discover/configure
+    a bearer token (scitoken) or an X.509 credential, with options to
+    require/disable either of those, or all authentication entirely.
+
+    {parameters}
+    """
+    def __init__(
+        self,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        cert=None,
+        auth=None,
+        url=None,
+        force_noauth=False,
+        fail_if_noauth=False,
+        **kwargs,
+    ):
+        # initialise session
+        super().__init__(**kwargs)
+
+        # handle options
+        if force_noauth and fail_if_noauth:
+            raise ValueError(
+                "cannot select both force_noauth and fail_if_noauth",
+            )
+        if force_noauth:
+            return
+
+        # find creds if we can
+        _auth = self._init_auth(
+            auth=auth,
+            cert=cert,
+            token=token,
+            token_audience=token_audience,
+            token_scope=token_scope,
+            url=url,
+        )
+
+        # if no auth was found, and we need it, fail here
+        if not _auth and fail_if_noauth:
+            raise IgwnAuthError("no valid authorisation credentials found")
+
+    def _init_auth(
+        self,
+        auth=None,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        url=None,
+    ):
+        # -- user-provided objects
+        # the user gave us something specific, so always use it
+        dummy = (None, False, True)
+        if token not in dummy:
```

Comment:
What about falsey values like 0?

Reply:
I can only be bothered to protect the user from themselves so much. In this context I don't think confusing 0 for false is a problem, given that this isn't a valid token value, but the point is well taken.